### PR TITLE
Add Out_channel STM tests

### LINKED ISSUE #378: Long Out_channel shrinking runs cause timeout
Adding the missing bytes-involving bindings introduced a sister-issue to #373 albeit for `Out_channel`.
This was observed by @shym in https://github.com/ocaml-multicore/multicoretests/issues/373#issuecomment-1630881135
I've also observed it on #376:

For Cygwin trunk/5.2 part1 
https://github.com/ocaml-multicore/multicoretests/actions/runs/5553343778/jobs/10141775211
```
random seed: 186839348
generated error fail pass / total     time test name

[ ]    0    0    0    0 / 1000     0.0s Lin DSL In_channel test with Domain
[ ]    0    0    0    0 / 1000     0.0s Lin DSL In_channel test with Domain (generating)
[ ]   18    0    0   18 / 1000    62.9s Lin DSL In_channel test with Domain
[✓]   20    0    1   19 / 1000   107.3s Lin DSL In_channel test with Domain

[ ]    0    0    0    0 / 5000     0.0s Lin DSL Out_channel test with Domain
[ ]    4    0    0    4 / 5000    17.7s Lin DSL Out_channel test with Domain
[ ]   17    0    0   17 / 5000    89.4s Lin DSL Out_channel test with Domain
[ ]   30    0    0   30 / 5000   152.8s Lin DSL Out_channel test with Domain
[ ]   37    0    0   37 / 5000   213.0s Lin DSL Out_channel test with Domain (shrinking:   11)
[ ]   37    0    0   37 / 5000   275.9s Lin DSL Out_channel test with Domain (shrinking:   15.0005)
[ ]   37    0    0   37 / 5000   341.2s Lin DSL Out_channel test with Domain (shrinking:   18.0004)
[ ]   37    0    0   37 / 5000   407.6s Lin DSL Out_channel test with Domain (shrinking:   20.0004)
[ ]   37    0    0   37 / 5000   469.6s Lin DSL Out_channel test with Domain (shrinking:   21.0007)
[ ]   37    0    0   37 / 5000   533.5s Lin DSL Out_channel test with Domain (shrinking:   22.0007)
[ ]   37    0    0   37 / 5000   598.2s Lin DSL Out_channel test with Domain (shrinking:   23.0007)
[ ]   37    0    0   37 / 5000   663.0s Lin DSL Out_channel test with Domain (shrinking:   24.0006)
...
[ ]   37    0    0   37 / 5000  3898.2s Lin DSL Out_channel test with Domain (shrinking:   73.0009)
[ ]   37    0    0   37 / 5000  3963.1s Lin DSL Out_channel test with Domain (shrinking:   74.0008)
[ ]   37    0    0   37 / 5000  4023.3s Lin DSL Out_channel test with Domain (shrinking:   75.0007)
[ ]   37    0    0   37 / 5000  4085.3s Lin DSL Out_channel test with Domain (shrinking:   76.0007)
[ ]   37    0    0   37 / 5000  4147.7s Lin DSL Out_channel test with Domain (shrinking:   77.0007)
[ ]   37    0    0   37 / 5000  4210.4s Lin DSL Out_channel test with Domain (shrinking:   78.0007)
[ ]   37    0    0   37 / 5000  4276.4s Lin DSL Out_channel test with Domain (shrinking:   79.0006)
[ ]   37    0    0   37 / 5000  4339.0s Lin DSL Out_channel test with Domain (shrinking:   80.0005)
[ ]   37    0    0   37 / 5000  4399.3s Lin DSL Out_channel test with Domain (shrinking:   81.0004)
[ ]   37    0    0   37 / 5000  4466.2s Lin DSL Out_channel test with Domain (shrinking:   82.0004)
[ ]   37    0    0   37 / 5000  4528.5s Lin DSL Out_channel test with Domain (shrinking:   83.0003)
Error: The operation was canceled.
```

For Windows MingW 5.1
https://github.com/ocaml-multicore/multicoretests/actions/runs/5553343781/jobs/10141774703
```
random seed: 163892320
generated error fail pass / total     time test name

[ ]    0    0    0    0 / 1000     0.0s Lin DSL In_channel test with Domain
[ ]    0    0    0    0 / 1000     0.0s Lin DSL In_channel test with Domain (generating)
[ ]    1    0    0    1 / 1000    63.0s Lin DSL In_channel test with Domain (shrinking:   29.0006)
[✓]    2    0    1    1 / 1000   108.9s Lin DSL In_channel test with Domain

[ ]    0    0    0    0 / 5000     0.0s Lin DSL Out_channel test with Domain
[ ]   11    0    0   11 / 5000    21.4s Lin DSL Out_channel test with Domain
[ ]   11    0    0   11 / 5000    86.6s Lin DSL Out_channel test with Domain (shrinking:    3.0004)
[ ]   11    0    0   11 / 5000   148.9s Lin DSL Out_channel test with Domain (shrinking:    5.0009)
[ ]   11    0    0   11 / 5000   212.4s Lin DSL Out_channel test with Domain (shrinking:    7.0010)
[ ]   11    0    0   11 / 5000   273.5s Lin DSL Out_channel test with Domain (shrinking:    8.0005)
[ ]   11    0    0   11 / 5000   336.8s Lin DSL Out_channel test with Domain (shrinking:    9.0011)
[ ]   11    0    0   11 / 5000   401.0s Lin DSL Out_channel test with Domain (shrinking:   12.0007)
....
[ ]   11    0    0   11 / 5000 10347.9s Lin DSL Out_channel test with Domain (shrinking:   87.0033)
[ ]   11    0    0   11 / 5000 10409.4s Lin DSL Out_channel test with Domain (shrinking:   88.0009)
[ ]   11    0    0   11 / 5000 10472.8s Lin DSL Out_channel test with Domain (shrinking:   88.0014)
[ ]   11    0    0   11 / 5000 10532.9s Lin DSL Out_channel test with Domain (shrinking:   89.0007)
[ ]   11    0    0   11 / 5000 10601.2s Lin DSL Out_channel test with Domain (shrinking:   89.0014)
[ ]   11    0    0   11 / 5000 10670.7s Lin DSL Out_channel test with Domain (shrinking:   90.0009)
[ ]   11    0    0   11 / 5000 10735.0s Lin DSL Out_channel test with Domain (shrinking:   90.0016)
[ ]   11    0    0   11 / 5000 10797.9s Lin DSL Out_channel test with Domain (shrinking:   91.0013)
[ ]   11    0    0   11 / 5000 10862.2s Lin DSL Out_channel test with Domain (shrinking:   91.0020)
[ ]   11    0    0   11 / 5000 10928.2s Lin DSL Out_channel test with Domain (shrinking:   91.0028)
[ ]   11    0    0   11 / 5000 10994.4s Lin DSL Out_channel test with Domain (shrinking:   92.0012)
[ ]   11    0    0   11 / 5000 11057.5s Lin DSL Out_channel test with Domain (shrinking:   93.0009)
[ ]   11    0    0   11 / 5000 11117.7s Lin DSL Out_channel test with Domain (shrinking:   93.0015)
[ ]   11    0    0   11 / 5000 11180.9s Lin DSL Out_channel test with Domain (shrinking:   94.0003)
[ ]   11    0    0   11 / 5000 11242.1s Lin DSL Out_channel test with Domain (shrinking:  102.0003)
[ ]   11    0    0   11 / 5000 11302.1s Lin DSL Out_channel test with Domain (shrinking:  108)
[✓]   12    0    1   11 / 5000 11331.5s Lin DSL Out_channel test with Domain

....
                                                              |                          
                                     Out_channel.output_substring t "bceY\145CQx\156" 1 6
                                                              |                          
                                   .-----------------------------------------------------.
                                   |                                                     |                          
                       Out_channel.close_noerr t                                Out_channel.length t                


+++ Messages ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

Messages for test Lin DSL Out_channel test with Domain:

  Results incompatible with sequential execution

                                                                   |                               
                                     Out_channel.output_substring t "bceY\145CQx\156" 1 6 : Ok (())
                                                                   |                               
                                   .---------------------------------------------------------------.
                                   |                                                               |                               
                     Out_channel.close_noerr t : ()                                  Out_channel.length t : Ok (6)                 

================================================================================
success (ran 2 tests)
```

### PR DESCRIPTION
**Pull request by jmid**
The `Lin` `Out_channel` test is a sore point of the test suite:
- it required several revisions to get to the current state #387 #370 #319
- it takes too long to run on macOS and has therefore been disabled #326
- it doesn't always trigger on FreeBSD #401
- it occasionally takes very long to run or shrink which can cause timeouts #378

`Out_channel`s buffer internally, meaning `length`'s result can vary.
This is a bad fit for `Lin`'s sequential consistency test, because we may end up finding counterexamples of different buffering (between a parallel and any interleaved sequential run) - or just spend a long time searching for one.
As such, a model-based `STM` test is a better fit - and this PR offers one.

The new model-based test
- tests sequential and parallel usage
- runs across all CI platforms as a positive test
- covers as much of the `Out_channel` module as the existing `Lin` test
- runs much more predictably
- enables us to test Out_channel on macOS again (without any issues)

Many things thus speak for switching to it.

On the other hand, despite its downsides the old Lin test has also managed to stress the runtime into triggering defects #412

Technically, there are a few nuggets in the test code:
- we model both closed and open `Out_channel`s with suitable preconditions. As such the test can generate open-close-open-close `cmd` sequences.
- the uncertainty of buffering affecting `length`'s output is handled by comparing less-or-equal-to the model's length
- binary-mode (only available on Windows) is modeled in the state
  - when enabled on MinGW/Cygwin it means `length` increases by 2 for every `'\n'` output (since `'\n'` maps to `"\r\n"`)
  - `position` does not reflect this translation though
  - the tests finally found a counterexample illustrating how buffering may interfere with `set_binary_mode`: it may not be the mode enabled at output-call-time that takes effect (example below). This is solved by always calling `flush` before `set_binary_mode` on MinGW/Cygwin where this matters:
   ```
   Open_text : Ok (())
   Set_binary_mode true : ()
   Output_char '\n' : Ok (())
   Set_binary_mode false : ()
   Flush : ()
   Length : 2
   ```

Closes #378 and #401

### CODE REVIEW

**Comment by jmid** [comment]
CI summary for 2150652:
- All 6 Cygwin/MinGW `trunk` workflows failed with a version mismatch, since `trunk` is now `5.3`
- linux-arm64-5.2 aborted `domain_spawntree` with `Fatal error: Failed to create domain` #428

Out of 59 workflows 7 failed with 1 genuine failure and 6 ci setup issues

**Comment by jmid** [comment]
I've rebased the PR on `main` after the merge of #429

**Comment by jmid** [comment]
CI summary for 024dd85
- 4 workflows aborted `domain_spawntree` with `Fatal error: Failed to create domain` #428
  - Cygwin 5.2 
  - Cygwin trunk 
  - FP trunk 
  - MinGW 5.2 
- MinGW bytecode 5.2 timed out during `threadomain` #203

Out of 44 workflows 5 failed with all 5 being genuine failures.

File: src/io/stm_tests.ml [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+    | Is_buffered -> cst0 "Is_buffered" fmt
+
+  let show_cmd = Util.Pp.to_show pp_cmd
+
+  (* a path and an open channel to that file; we need to keep the path
+     to cleanup after the test run *)
+  type sut = { path            : string;
+               mutable channel : Out_channel.t }
+
+  type state = Closed of int64
```
Comment (shym) [inline src/io/stm_tests.ml:53]: I don’t think I’ve seen the argument of `Closed` used anywhere. Did I miss it?

File: src/io/stm_tests.ml [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+        let position = Int64.add position (Int64.of_int l) in
+        let length = (* Windows text mode maps '\n' to "\r\n" *)
+          Int64.add length
+            (if (Sys.win32 || Sys.cygwin) && not binary_mode
+             then Int64.of_int (l + count_nls String.(sub str p l))
+             else Int64.of_int l) in
+        Open { position; length; buffered; binary_mode; }
+
+  let init_sut () =
+    let path = Filename.temp_file "lin-dsl-" "" in
```
Comment (shym) [inline src/io/stm_tests.ml:185]: It could be renamed as `"stm-"`, maybe?
There are a couple of occurrences of `lin-dsl-` in the code.

File: src/io/stm_tests.ml [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+        | Open { position = _; length; buffered = _; binary_mode = _ } -> i <= length)
+    | Close, Res ((Result (Unit,Exn),_), r) ->
+       (match s,r with
+         | Closed _, Error (Sys_error _) (*"Close exception" - unspecified *)
+         | Open _, Ok () -> true
+         | _ -> false)
+    | Close_noerr, Res ((Unit,_), r) ->
+       (match s,r with
+         | Closed _, ()
+         | Open _, () -> true)
```
Comment (shym) [inline src/io/stm_tests.ml:249]: This might be squashed to `true`.

File: src/io/stm_tests.ml [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+        Int64.add length
+          (if (Sys.win32 || Sys.cygwin) && not binary_mode && c = '\n' then 2L else 1L) in
+      Open { position; length; buffered; binary_mode; }
+    | Output_byte i, Open { position; length; buffered; binary_mode } ->
+      let position = Int64.succ position in
+      let length = (* Windows text mode maps '\n' to "\r\n" *)
+        Int64.add length
+          (if (Sys.win32 || Sys.cygwin) && not binary_mode && (i mod 256 = 10) then 2L else 1L) in
+      Open { position; length; buffered; binary_mode; }
+    | Output_string arg, Open { position; length; buffered; binary_mode } ->
```
Comment (shym) [inline src/io/stm_tests.ml:141]: I wondered there and on the following cases whether it could be possible to factor most of the logic out. I didn’t see an obviously good way to go, though.

File: src/io/stm_tests.ml [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+        | _ -> false)
+    | Seek _, Res ((Unit,_), ()) -> true
+    | Pos, Res ((Result (Int64,Exn),_), r) ->
+       (match s with
+        | Closed _ -> true (*r = Error (Invalid_argument "Pos exception") - unspecified *)
+        | Open { position; length = _; buffered = _; binary_mode = _ } -> r = Ok position)
+    | Length, Res ((Int64,_),i) ->
+       (match s with
+        | Closed _ -> true
+        | Open { position = _; length; buffered = _; binary_mode = _ } -> i <= length)
```
Comment (shym) [inline src/io/stm_tests.ml:240]: This is nice!
I wonder if it should be explicitly documented, as this is the reason of the negativity of Lin tests.

**Review by shym — APPROVED** [review]
A real test for `Out_channel`s, where we explicitly take care of the buffering, this is very nice! Congratulations!

**Comment by shym** [comment]
> On the other hand, despite its downsides the old Lin test has also managed to stress the runtime into triggering defects #412

Do you think the STM tests are less likely to trigger such defects?

File: src/io/stm_tests.ml (hunk 1 above)
Comment (jmid) [reply src/io/stm_tests.ml]: Indeed! I think this is a left-over from when I started. 
Initially I think I allowed all `cmd`s in both states and tried to track length for closed `Out_channel`s too.
Fixed in df541cc

File: src/io/stm_tests.ml (hunk 2 above)
Comment (jmid) [reply src/io/stm_tests.ml]: Indeed! Fixed in e168a46 and cleaned some more in 9c63209

File: src/io/stm_tests.ml (hunk 4 above)
Comment (jmid) [reply src/io/stm_tests.ml:144]: There is indeed a lot of repetition!
I suggest a simple refactoring in 9bd3253

File: src/io/stm_tests.ml (hunk 5 above)
Comment (jmid) [reply src/io/stm_tests.ml:240]: Good point. Added a comment to that effect in b247e41

**Comment by jmid** [comment]
Thanks for the review!

Discussing this PR with you triggered me to look up previous counterexamples reported by the `Lin Out_channel` test.
Indeed, the last merge to `main` a571972 triggered, e.g., the following for the Linux 5.2 debug workflow:
```
Messages for test Lin Out_channel test with Domain:

  Results incompatible with sequential execution

                                           |                  
                         Out_channel.output_byte t 2 : Ok (())
                                           |                  
                       .--------------------------------------.
                       |                                      |                  
         Out_channel.length t : Ok (1)         Out_channel.close_noerr t : ()   
```
This should have a possible sequentialization (below) which we seem to miss:
```
Out_channel.output_byte t 2;;
Out_channel.length t;;
Out_channel.close_noerr t;;
```

Looking at previous counterexamples also made me realize that
- some of them are triggered by using `close`/`close_noerr` in parallel. However we won't generate such inputs in this `STM` test version ATM unfortunately, because they don't have a well-balanced `open`-`close` serialization. I want to experiment with lifting this limitation.
- other `Lin Out_channel` counterexamples are triggered by weird output unlinearizable behaviours arising from running other `cmd`s on already closed channels. This is a gray-zone specification-wise, which does not offer an immediate benefit AFAICS. As such, I'm more reluctant to invest in lifting it.
  ```
                                                                                                                   |                                                       
                                                                                                     Out_channel.flush t : Ok (())                                         
                                                                 Out_channel.output_bytes t "7)\016)\170\245A{(2\255\000mr\185\243\206\2475-\192i\135\2216}w" : Ok (())    
                                                                                             Out_channel.set_binary_mode t false : Ok (())                                 
                                                                                                                   |                                                       
                                                           .---------------------------------------------------------------------------------------------------------------.
                                                           |                                                                                                               |                                                       
                                             Out_channel.length t : Ok (27)                                                                                  Out_channel.close_noerr t : ()                                        
                                         Out_channel.is_buffered t : Ok (true)                                                                               Out_channel.close t : Ok (())                                         
                                         Out_channel.output_byte t 6 : Ok (())                                                                               Out_channel.pos t : Ok (65536)                                        
                                                                                                                                                         Out_channel.is_buffered t : Ok (true)                                     
                                                                                                                                     Out_channel.output_bytes t "K\226" : Error (Sys_error("Bad file descriptor"))                 
                                                                                                                                                         Out_channel.is_buffered t : Ok (true)                                     
                                                                                                                     Out_channel.output_substring t "\197\180\2484g\180\230\193" 4 7 : Error (Invalid_argument("output_substring"))
                                                                                                                                                             Out_channel.flush t : Ok (())                                         
  ```

Our discussion also made me check my recollection regarding locking:
https://github.com/ocaml/ocaml/blob/cedff5854ac91dccf5847b527192223ef506b1e2/runtime/io.c#L60
```
All operations on channels first take the channel lock.
```
As such, they `Out_channel` operations should act atomically when used in parallel and be sequential consistent.


>> On the other hand, despite its downsides the old Lin test has also managed to stress the runtime into triggering defects #412

> Do you think the STM tests are less likely to trigger such defects?

Possibly, but I'm unsure.
- A single `Lin` input will call the `Out_channel` interface (up to) 50 times * 1 parallel run * #interleavings times.
- In comparison a single `STM` input will call the interface (up to) 25 times * 1 parallel run.

Since #interleavings can be quite high, a corner case bug is more likely to trigger more often when it is run more.
On the other hand, when considering this behaviour across a `count` of 1000, the negative `Lin` test will finish sooner (and report a "counterexample") whereas the current `STM` test runs to completion (1000 inputs * 25 repetitions).

**Comment by jmid** [comment]
In 8236876 I extend the `STM` tests to be able to generate `close` and `close_noexn` in state `Closed` too.

File: src/io/stm_tests.ml (hunk 3 above)
Comment (jmid) [reply src/io/stm_tests.ml]: This one I needed to change and extend in 8236876

**Comment by shym** [comment]
> This should have a possible sequentialization (below) which we seem to miss:

In the sequentialized version, the `length` will most probably be 0, as the buffer is not flushed; the output we see is due to the `close_noerr` being half-run, having flushed but not closed yet (as those two steps are explicitly not run atomically), isn’t it?

>  This is a gray-zone specification-wise, which does not offer an immediate benefit AFAICS.

To make sure that I understand correctly what this is doing, because I think I missed it in my first review: due to the `postcond` function, we’ll never test `close` or `close_noerr` in one of the parallel branches except when the other branch is only doing `open` and `close` calls, or am I mistaken in how `postcond` is used in `all_interleavings_ok`? This would mean those tests would never run into issues such as ocaml/ocaml#11878 I suppose?

**Comment by jmid** [comment]
> > This should have a possible sequentialization (below) which we seem to miss:
> 
> In the sequentialized version, the `length` will most probably be 0, as the buffer is not flushed; the output we see is due to the `close_noerr` being half-run, having flushed but not closed yet (as those two steps are explicitly not run atomically), isn’t it?

This is a probable explanation indeed! :+1: 
However it seems a bit unsatisfying that this stops testing and is reported as a counterexample of parallelism-safety.

> > This is a gray-zone specification-wise, which does not offer an immediate benefit AFAICS.
> 
> To make sure that I understand correctly what this is doing, because I think I missed it in my first review: due to the `postcond` function, we’ll never test `close` or `close_noerr` in one of the parallel branches except when the other branch is only doing `open` and `close` calls, or am I mistaken in how `postcond` is used in `all_interleavings_ok`? This would mean those tests would never run into issues such as [ocaml/ocaml#11878](https://github.com/ocaml/ocaml/issues/11878) I suppose?

Point well taken - and yes, you are right :+1: 
Following the Erlang version described in the ICFP'09 paper, `STM` will generate candidate "Y"-triples and only keep those which satisfy all `precond`itions in all interleavings (which is what `all_interleavings_ok` should express). This indeed leaves out two parallel `close*` `cmd`s.

I've therefore taken a pass over the `STM` test and extended the `cmd`s so that all of them can also occur in state `Closed`. The only generator and `precond` limitation now is that `Open_text` is not allowed in state `Open` (this was also a limitation of the `Lin` test).

This revealed a divergence from the spec as I shared offline:
```ocaml
val close : t -> unit
(** Close the given channel, flushing all buffered write operations.  Output
    functions raise a [Sys_error] exception when they are applied to a closed
    output channel, except {!close} and {!flush}, which do nothing when applied
    to an already closed channel.  Note that {!close} may raise [Sys_error] if
    the operating system signals an error when flushing or closing. *)
```
Yet `output_string`, `output_bytes`, `output`, and `output_substring` of length 0 on a `Closed` channel does not fail, e.g.:
```
--- Failure --------------------------------------------------------------------

Test STM Out_channel test sequential failed (22 shrink steps):

   Close
   Output_string ""


+++ Messages ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

Messages for test STM Out_channel test sequential:

  Results incompatible with model

   Close : Ok (())
   Output_string "" : Ok (())
```

For now, I've adjust `postcond` to accept these cornercases.
We should consider submitting a PR to adjust the specification though...

**Comment by jmid** [comment]
Some overdue CI summaries...

For 9c63209:
- Cygwin trunk aborted during `domain_spawntree` #428
- MinGW bytecode 5.1 crashed during `threadomain` #203
- MinGW bytecode 5.2 aborted during `domain_spawntree` #428
- MinGW trunk aborted during `domain_spawntree` #428
- macOS trunk failed to build `dune` #433
- Cygwin 5.2 was cancelled after 138m during Dynlink #307

Out of 44 workflows 5 failed and 1 was cancelled, all 6 due to genuine issues


For 8236876:
- Cygwin 5.2 aborted during `domain_spawntree` #428
- Cygwin trunk aborted during `domain_spawntree` #428
- MinGW 5.2 aborted during `domain_spawntree` #428
- MinGW bytecode trunk aborted during `domain_spawntree` #428
- macOS trunk failed to build `dune` #433

Out of 44 workflows 5 failed with all 5 being genuine issues


For 0c8fdcb:
- MinGW bytecode 5.1 timed out in `threadomain` #203
- 4 workflows aborted during `domain_spawntree` #428
  - Cygwin 5.2
  - Cygwin trunk
  - MinGW 5.2
  - macOS 5.2
- macOS trunk failed to install dune #433
- MinGW 5.1 failed the new `STM` `Out_channel` test with `flush` raising `Sys_error("Bad file descriptor")` when used in parallel
- all 21 5.2 and trunk/5.3 workflows failed the new `STM` `Out_channel` due to the exception regression #432 
  - 32bit 5.2
  - 32bit trunk
  - Bytecode 5.2
  - Bytecode trunk
  - Cygwin 5.2
  - Cygwin trunk
  - FP 5.2
  - FP trunk
  - Linux 5.2
  - Linux 5.2 debug
  - Linux trunk
  - Linux trunk debug
  - MinGW 5.2
  - MinGW bytecode 5.2
  - MinGW bytecode trunk
  - MinGW trunk
  - macOS 5.2
  - linux-arm64-5.2
  - linux-s390x-5.2
  - linux-ppc64le-5.2
  - macos-arm64-5.2

Out of 44 workflows 24 failed with a total of 28 failures (4 workflows failed two tests) with all of them being genuine

**Comment by jmid** [comment]
The PR's latest test revision is so effective at triggering the #432 regression that we should consider temporarily relaxing the test to accept the current behaviour, to avoid having to check ~20 failing workflows on each CI run... :grimacing:

**Review by shym — APPROVED** [review]
Looks very good, thank you for the updates!

**Comment by jmid** [comment]
I rebased on `main` to avoid more false alarms and added 86f2ab2 to temporarily disable the #432 regression reports.

**Comment by jmid** [comment]
CI summary for merge to `main`:
- macOS trunk failed to build `dune` #433

Out of 45 workflows 1 failed with a genuine issue
